sdk/servicebus/azure-servicebus/azure/servicebus/_common/auto_lock_renewer.py

```diff
@@ -0,0 +1,132 @@
+            :language: python
+            :dedent: 4
+            :caption: Automatically renew a session lock
+
+    """
+
+    def __init__(self, executor=None, max_workers=None):
+        self.executor = executor or ThreadPoolExecutor(max_workers=max_workers)
+        self._shutdown = threading.Event()
+        self.sleep_time = 1
+        self.renew_period = 10
+
+    def __enter__(self):
+        if self._shutdown.is_set():
+            raise ServiceBusError("The AutoLockRenew has already been shutdown. Please create a new instance for"
+                                  " auto lock renewing.")
+        return self
+
+    def __exit__(self, *args):
+        self.shutdown()
+
+    def _renewable(self, renewable):
+        if self._shutdown.is_set():
+            return False
+        if hasattr(renewable, 'settled') and renewable.settled:
+            return False
+        if isinstance(renewable, ServiceBusSession) and not renewable._receiver._running:
+            return False
+        if renewable.expired:
+            return False
+        return True
+
+    def _auto_lock_renew(self, renewable, starttime, timeout, on_lock_renew_failure=None):
+        _log.debug("Running lock auto-renew thread for %r seconds", timeout)
+        error = None
+        try:
+            while self._renewable(renewable):
+                if (utc_now() - starttime) >= datetime.timedelta(seconds=timeout):
+                    _log.debug("Reached auto lock renew timeout - letting lock expire.")
+                    raise AutoLockRenewTimeout("Auto-renew period ({} seconds) elapsed.".format(timeout))
+                if (renewable.locked_until_utc - utc_now()) <= datetime.timedelta(seconds=self.renew_period):
+                    _log.debug("%r seconds or less until lock expires - auto renewing.", self.renew_period)
+                    renewable.renew_lock()
+                time.sleep(self.sleep_time)
+        except AutoLockRenewTimeout as e:
+            renewable.auto_renew_error = e
+        except Exception as e:  # pylint: disable=broad-except
+            _log.debug("Failed to auto-renew lock: %r. Closing thread.", e)
+            error = AutoLockRenewFailed(
+                "Failed to auto-renew lock",
+                inner_exception=e)
+            renewable.auto_renew_error = error
+        finally:
+            if on_lock_renew_failure:
+                if self._shutdown.is_set() \
+                        or (hasattr(renewable, 'settled') and renewable.settled) \
+                        or (isinstance(renewable, ServiceBusSession) and not renewable._receiver._running) \
+                        or (not error and not renewable.expired):
+                    # Basically we want to make sure that any of the "intentional" exit cases are not fired on.
+                    # e.g. shutdown, settlement, session closure, renewer timeout (as opposed to renew failure/expiry)
+                    return
```

Comment:
I think the code could be simplified to:
```
if renewable.auto_renew_error and self._renewable(renewable):  # error happened while the message/session is still renewable.
    on_lock_renew_failure(renewable)
```

Reply:
That haunted me this whole time because it's _very similar_ but some of the renewable conditions are things we explicitly DON'T want to care about since they're "intentional".  When I first wrote this clause it was only 1-2 conditions and seemed proper to have them be special-cased, but it grew since I made that decision, so I'll reconsider it again.

FYI I very intentionally did NOT use auto_renew_error (and did the local variable "Error" instead) because technically the auto_renew_error property could be from a stale instantiation of an autorenewer.  Very very unlikely, but possible.  (this doesn't invalidate the core of your point, just wanted to give my thinking since it was part of your suggestion)

Reply:
I'm pretty sure we can't safely replace the clause as written.  (It took me way too long of staring at this to come to that conclusion, so I'm both ashamed with myself and lacking in confidence, so don't hesitate to second guess)

The problem at hand is renewable.expired.  _renewable() returns false if expired is true, but that is not sufficient to say if the callback should or should not run; e.g. if the renew_period is set too large such that the message/session expires "naturally", I'd expect the callback to execute because it was an "unexpected" lock-loss (even though due to a configuration the user could set.)

I'm pretty sure there's also incongruity in how the final check re: "not error and not expired" carried over as well but I'm struggling to mentally summarize the decision tree effectively; and am going to for now err towards using the more clear set of conditions explicitly even though it duplicates some logic.  I'll leave a note here with your intuition, as I think it's on the right track, just that there's this subtlety with how "renewable in the midst" vs "renewable at the end of things" plays out and trying to coerce the former to the latter adds more complexity than needed.

Reply:
Thanks to Richard indulging my contemplating of refactoring options, revisited this with a "third option" and converted the conditional to use a flag variable that far better captures the state as needed.